VSS.md

```diff
@@ -0,0 +1,518 @@
+- **search\_hnsw\_create\_exceptions\_count**	(Integer)	Count of HNSW creation exceptions.  
+- **search\_hnsw\_search\_exceptions\_count**	(Integer)	Count of HNSW search exceptions  
+- **search\_hnsw\_remove\_exceptions\_count**	(Integer)	Count of HNSW removal exceptions.  
+- **search\_hnsw\_add\_exceptions\_count**	(Integer)	Count of HNSW addition exceptions.  
+- **search\_hnsw\_modify\_exceptions\_count**	(Integer)	Count of HNSW modification exceptions  
+- **search\_modify\_subscription\_skipped\_count**	(Integer)	Count of skipped subscription modifications  
+- **search\_remove\_subscription\_successful\_count**	(Integer)	Count of successful subscription removals  
+- **search\_remove\_subscription\_skipped\_count**	(Integer)	Count of skipped subscription removals  
+- **search\_remove\_subscription\_failure\_count**	(Integer)	Count of failed subscription removals  
+- **search\_add\_subscription\_successful\_count**	(Integer)	Count of successfully added subscriptions  
+- **search\_add\_subscription\_failure\_count**	(Integer)	Count of failures of adding subscriptions  
+- **search\_add\_subscription\_skipped\_count**	(Integer)	Count of skipped subscription adding processes  
+- **search\_modify\_subscription\_failure\_count**	(Integer)	Count of failed subscription modifications  
+- **search\_modify\_subscription\_successful\_count**	(Integer)	Count of successful subscription modifications
+
+### Key Functionality Gaps Relative to RediSearch 
+
+ValkeySearch's current focus is primarily on vector search, leveraging a robust foundation designed for extensibility. It is highly optimized for performance, delivering nearly linear scaling with an increasing number of CPU cores. Furthermore, ValkeySearch employs advanced techniques such as deduplication and string interning to maintain a low memory footprint. However, RediSearch, developed over several years, offers a more comprehensive feature set. The following outlines the key gaps:  
+
+**Unsupported Index Types**
+- Full-Text: Allows fast querying of textual data, supporting features such as  exact phrase matching, prefix matching, fuzzy matching and more. 
+- Geospatial: Allows storing, querying, and manipulating location-based data such as radius queries, nearest neighbor searches, and distance calculations, leveraging sorted sets and geohashes.
+
+**Functionality Gaps**
+- Non-vector queries: Currently, non-vector indexes are utilized only within the context of vector search queries. Expanding support to standalone non-vector use cases broadens applicability. For instance, a query could search for products priced within a specific range, tagged as "sport," and with a title containing the word "basketball." An example syntax for such a query might look like:  
+
+```
+FT.SEARCH index_name "@price:[min max] @tags:{sport} @title:basketball"
+```
+
+- ACLs: allowing query and index operations to respect the permissions configured for each user.
```

Comment:
Is this about the ACL rules for key patterns?

What's the idea here? Should users be able to search only for keys they have access to read?

Technically, how could this work? Search all keys first and then filter before returning the results to the client?

Reply:
The security model that AWS implemented directly supports the prefix-based definition of indexes. This implementation operates at the start of a command and checks if the current user has read access to any of the keys that MIGHT be in the index, i.e., current user must have read access to 100% of the keyspace covered by the PREFIX clause of the creation of the index. Once that check passes, no additional security check would be performed.

Reply:
hmm, I think that we should discuss this. I do have some reservations about this approach because ACLs might be changed after an index is created and ft.search should still honor the ACLs.

Reply:
Perhaps my explanation was poorly worded. The ACL checks operate on every command. My comment about "no additional security check would be performed" meant to say that there's no key-level access checking. An individual user either has 100% or 0% access to an index which is determined right at the start of the command.